File: modules/gui.control/src/macos/label.m
#import <Cocoa/Cocoa.h>

#include <gui.control/label.h>
#include <gui.platform.macos/gui.platform.macos.h>

static NSView* mel__label_construct(Mel_Gui_Handle h, const Mel_Gui_Create_Desc* desc)
{
    (void)h;
    NSTextField* tf = [[NSTextField alloc] initWithFrame:NSMakeRect(0, 0, desc->w, desc->h)];
    [tf setEditable:NO];
    [tf setBordered:NO];
    [tf setBezeled:NO];
    [tf setDrawsBackground:NO];
    [tf setSelectable:NO];

    if (desc->id == 1) {
        [tf setFont:[NSFont systemFontOfSize:22 weight:NSFontWeightSemibold]];
    }

    NSString* s = [[NSString alloc] initWithBytes:desc->text.data
                                           length:(NSUInteger)desc->text.len
                                         encoding:NSUTF8StringEncoding];
    [tf setStringValue:(s ?: @"")];
    return tf;
}

void mel_gui_label_platform_register(Mel_Atom atom)
{
    mel_gui_macos_register_constructor(atom, mel__label_construct);
}

File: modules/gui.control/src/win32/label.c
#include <gui.control/label.h>
#include <gui.platform.win32/gui.platform.win32.h>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>

static HWND mel__label_construct(Mel_Gui_Handle h, const Mel_Gui_Create_Desc* desc, HWND parent_hwnd)
{
    (void)h;
    wchar_t* text = mel_gui_win32_str8_to_wide(desc->text);
    HWND hwnd = CreateWindowExW(
        0, L"STATIC", text ? text : L"",
        WS_CHILD | WS_VISIBLE | SS_LEFT,
        0, 0, desc->w, desc->h,
        parent_hwnd, (HMENU)(UINT_PTR)desc->id,
        mel_gui_win32_hinstance(), NULL);
    mel_gui_win32_wide_free(text);
    return hwnd;
}

void mel_gui_label_platform_register(Mel_Atom atom)
{
    mel_gui_win32_register_constructor(atom, mel__label_construct);
}
